### PR TITLE
asm: complete arm64 instruction encodings for homemade assembler

### LINKED ISSUE #233: JIT: replace golang-asm with a thread-safe one (from scratch)
In aarch64, golang-asm is not thread-safe, and [the multi-goroutine engine test case](https://github.com/tetratelabs/wazero/blob/0186e41b27e572964e9f018cd9cd8a655eb6d3f8/tests/engine/adhoc_test.go#L67-L87) actually fails if gorutine num > 1.

We could introduce a mutex in engine.Compile and lock the golang-asm usage, then the test passes, but it's not ideal.

Instead, we should replace golang-asm with a thread-safe, but there's not alternative out there. 

That means we have to implement an assembler from scratch.

### PR DESCRIPTION
**Pull request by mathetake**
This commit implements all the arm64 instruction encodings necessary 
for our JIT compiler and replaces the golang-asm assembler with our
handmade assembler on arm64 platform. Notably, this allows us to do 
concurrent compilations. 

This closes #233 combined with #406.

### CODE REVIEW

**Comment by mathetake** [comment]
ok progress: 30%

**Comment by mathetake** [comment]
OK the left is Memory <-> Register data transfers, meaning that I have to implement addressing, the hardest part :D

**Comment by mathetake** [comment]
95% of encoding done.. the left is relative jumps, and then we could run e2e tests

**Comment by mathetake** [comment]
ok finished all encodings.. debugging time to match the binary output between homemade and golang-asm

**Comment by mathetake** [comment]
ok  turns out that the constant pool strategy differs from Go's assembler.. will do next Monday..

**Comment by mathetake** [comment]
almost there :D

**Comment by mathetake** [comment]
ok passed all tests and the output binary 100% matches the ones from golang-asm  👯  will polish in the next few hours and then good to go!
!img[https://media.giphy.com/media/26BnaOFdD5Xbo1RMQ/giphy.gif]

**Comment by mathetake** [comment]
ok ready for review!

File: internal/wasm/jit/engine.go [hunk 1]
```diff
@@ -786,14 +786,7 @@ func (ce *callEngine) builtinFunctionMemoryGrow(mem *wasm.MemoryInstance) {
 	ce.moduleContext.memoryElement0Address = bufSliceHeader.Data
 }
 
-// golang-asm is not goroutine-safe so we take lock until we complete the compilation.
-// TODO: delete after https://github.com/tetratelabs/wazero/issues/233
-var assemblerMutex = &sync.Mutex{}
```
Comment (mathetake) [inline internal/wasm/jit/engine.go:791]: 🕺

File: go.mod [hunk 2]
```diff
@@ -4,8 +4,9 @@ module github.com/tetratelabs/wazero
 go 1.17
 
 require (
+	// Test-only dependency.
 	github.com/stretchr/testify v1.7.0
-	// Once we reach some maturity, remove this dep and implement our own assembler.
+	// Test-only dependency.
```
Comment (mathetake) [inline go.mod:9]: 💃

File: internal/asm/arm64/impl.go [hunk 3]
```diff
@@ -0,0 +1,2340 @@
...
+	code := a.Bytes()
+	for _, cb := range a.OnGenerateCallbacks {
+		if err := cb(code); err != nil {
+			return nil, err
+		}
+	}
+	return code, nil
+}
+
+// maybeFlushConstPool maybe flushes the current constants in the constPool.
```
Comment (codefromthecrypt) [inline internal/asm/arm64/impl.go:268]: maybeFlushConstPool flushes the constant pool if endOfBinary or a boundary condition was met.

File: internal/asm/arm64/impl.go [hunk 4]
```diff
@@ -0,0 +1,2340 @@
...
+		err = a.EncodeSIMDByteToRegister(n)
+	case OperandTypesTwoSIMDBytesToSIMDByteRegister:
+		err = a.EncodeTwoSIMDBytesToSIMDByteRegister(n)
+	default:
+		err = fmt.Errorf("encoder undefined for [%s] operand type", n.Types)
+	}
+	return
+}
+
+// CompileStandAlone implements asm.AssemblerBase.CompileStandAlone
```
Comment (codefromthecrypt) [inline internal/asm/arm64/impl.go:379]: nested linking doesn't seem to navigate, which is why I have been switching to the following when implementing something out of package.

```
Xxx implements the same method as documented on yyy.Zzzz.
```

File: internal/asm/arm64/impl.go [hunk 5]
```diff
@@ -0,0 +1,2340 @@
...
+	n := a.newNode(CSET, OperandTypesRegisterToRegister)
+	n.SrcReg = conditionalRegisterStateToRegister(cond)
+	n.DstReg = dstReg
+}
+
+func errorEncodingUnsupported(n *NodeImpl) error {
+	return fmt.Errorf("%s is unsupported for %s type", InstructionName(n.Instruction), n.Types)
+}
+
+func (a *AssemblerImpl) EncodeNoneToNone(n *NodeImpl) (err error) {
```
Comment (codefromthecrypt) [inline internal/asm/arm64/impl.go:528]: need to be exported? if so // EncodeNoneToNone is exported because XXX

File: internal/asm/arm64/impl.go [hunk 6]
```diff
@@ -0,0 +1,2340 @@
...
+	if n.Instruction != NOP {
+		err = errorEncodingUnsupported(n)
+	}
+	return
+}
+
+func (a *AssemblerImpl) EncodeJumpToRegister(n *NodeImpl) (err error) {
```
Comment (codefromthecrypt) [inline internal/asm/arm64/impl.go:535]: same on export

**Comment by codefromthecrypt** [comment]
I'm going to make a pass to do the little doc nits and push a commit vs have you do them :)

**Comment by codefromthecrypt** [comment]
a couple things maybe can be changed now (or a later PR)?

```go
// Note: this interface is coupled and heavily influenced by golang-asm's API (i.e. Go's official assembler).
// Therefore, we will do the refactoring after golang-asm removal.
type AssemblerBase interface {
--snip

	// BuildJumpTable calculates the offsets between the first instruction `initialInstructions[0]`
	// and others (e.g. initialInstructions[3]), and wrote the calculated offsets into pre-allocated
	// `table` slice in little endian.
	//
	// TODO: This can be hidden into assembler implementation after golang-asm removal.
	BuildJumpTable(table []byte, initialInstructions []Node)
```

**Comment by codefromthecrypt** [comment]
another TODO?
```
	// CompileJumpToMemory adds jump-type instruction whose destination is stored in the memory address specified by `baseReg`,
	// and returns the corresponding Node in the assembled linked list.
	//
	// Note: this has exactly the same implementation as the same method in asm.AssemblerBase in the homemade assembler.
	// Therefore, this will be removed after golang-asm removal.
	CompileJumpToMemory(jmpInstruction asm.Instruction, baseReg asm.Register)
```

**Comment by codefromthecrypt** [comment]
```
// Bytes returns the encoded binary.
func (a *AssemblerImpl) Bytes() []byte {
	// 16 bytes alignment to match our impl with golang-asm.
	// https://github.com/golang/go/blob/release-branch.go1.15/src/cmd/internal/obj/arm64/asm7.go#L62
	//
	// TODO: Delete after golang-asm removal.
	if pad := 16 - a.Buf.Len()%16; pad > 0 && pad != 16 {
		a.Buf.Write(make([]byte, pad))
	}
	return a.Buf.Bytes()
}
```

**Comment by mathetake** [comment]
I will resolve these todos later after I’m feeling comfortable to completely delete golang-ask code since resolving them unable me to keep golang-ask working

**Review by codefromthecrypt — APPROVED** [review]
OK I looked through all the changes and some similar things from the amd64 impl as well. 

Besides adding the RATIONALE including why we tend towards being like go assembly when in doubt (for familiarity and help troubleshooting), one thing to keep in mind is right now, we are depending heavily on round-tripping through the "debug" assemblers in order to get decent test coverage.

We shouldn't hunt coverage now, as I noticed a lot of code is just there to interop with go's assembler, and that would be pointless busywork. However, sometime between now and when that's removed, let's backfill some coverage in the natural unit tests for the parts that will stay, especially anything tricky. One way to prioritize could be to make comments saying "TODO: especially unit test this part" for places you lost time on in the past.

Besides that advice for post-merge, I'm really happy with the progress here and looking forward to the merge when you feel ready!

**Comment by mathetake** [comment]
Thanks for the help!!

**Pull request merged**